ReadMe.md
## Wes Anderson palettes for javascript

Color palettes based on Wes Anderson movies, taken from
[Karthik Ram](http://inundata.org/)'s
[wesanderson](https://github.com/karthik/wesanderson) package for R,
the first round of palettes of which were derived from
[the Wes Anderson Palettes Tumblr](http://wesandersonpalettes.tumblr.com).

![wes anderson palettes](palettes.png)

### Installation

You could grab the files directly from the github repository:
[`wesanderson.js`](https://raw.githubusercontent.com/kbroman/wesandersonJS/master/wesanderson.js)
and
[`wesanderson.css`](https://raw.githubusercontent.com/kbroman/wesandersonJS/master/wesanderson.css).

Or you could clone the git repository:

```
git clone git://github.com/kbroman/wesandersonJS
```

Or, you can use [bower](http://bower.io/):

```
bower install wesanderson
```

You'll then find the `wesanderson.js` file in
`bower_components/wesanderson/wesanderson.js`.

### Usage

#### Javascript palettes

[`wesanderson.js`](https://github.com/kbroman/wesandersonJS/blob/master/wesanderson.js)
contains two things:

- `wes_palettes` &ndash; an object indexed by palette name, containing
  arrays of colors. For example, `wes_palettes["Rushmore1"]` gives an
  array of 5 colors from the movie
  [Rushmore](http://en.wikipedia.org/wiki/Rushmore_%28film%29).

- `wes_palette` &ndash; a function to be called with a color palette name
  and (optionally) some number of colors. For example,
  `wes_palette("Rushmore1", 2)` will give you the first two colors in
  the `Rushmore1` palette.

#### CSS

[`wesanderson.css`](https://github.com/kbroman/wesandersonJS/blob/master/wesanderson.css)
is organized like the
[`colorbrewer.css`](https://github.com/jeanlauliac/colorbrewer/blob/master/colorbrewer.css)
file that was distributed with [D3](http://d3js.org) version 3. For each palette, it defines a set
of classes like:

```css
.Rushmore1.q0{fill:rgb(225,189,109)}
.Rushmore1.q1{fill:rgb(234,190,148)}
.Rushmore1.q2{fill:rgb(11,119,94)}
.Rushmore1.q3{fill:rgb(53,39,74)}
.Rushmore1.q4{fill:rgb(242,48,15)}
```

Assign classes like `q0`, `q1`, `q2`, ..., to the elements you wish to
be filled, and then set a class like `Rushmore1` to a parent element,
such as the SVG.

### License

Licensed under the [MIT license](License.md). ([More information here](http://en.wikipedia.org/wiki/MIT_License).)


NEWS.md
## wesandersonJS 0.7.1 (2018-05-01)

### New features

- Updated palettes to follow changes in
  [Karthik Ram](http://inundata.org/)'s
  [wesanderson R package](https://github.com/karthik/wesanderson).
  In particular, two new palettes for Isle of Dogs.


## wesandersonJS 0.6.3 (2017-05-07)

### New features

- Updated palettes to follow to changes in
  [Karthik Ram](http://inundata.org/)'s
  [wesanderson R package](https://github.com/karthik/wesanderson).

- Removed the scales; [d3](https://d3js.org) version 4 is now just
  using arrays of colors with more general scale functions.

### Minor changes

- Updated the tests to use [d3](https://d3js.org) version 4



## wesandersonJS 0.5.1 (2015-03-08)

### New features

- Added [d3](https://d3js.org) scales



## wesandersonJS 0.4.1 (2015-03-07)

### New features

- Added `wesanderson.css`.



## wesandersonJS 0.3.8 (2015-03-06)

### New features

- Added `wes_palette` function.



## wesandersonJS 0.3.3 (2015-03-06)

Initial release


test/grab_png.R
# grab image with palettes

if(!require(webshot)) {
    install.packages("webshot")
    library(webshot)
    webshot::install_phantomjs()
}

webshot("index.html", "palettes.png", selector="div#chart")
system("convert -crop 750x400+0+0 palettes.png ../palettes.png")


test/test_wesanderson.js
// Generated by CoffeeScript 2.2.4
// visualize the wes anderson palettes
var bgrect, dark_colors, gap_betw_cols, has_dark, height, hoffset, hscale, indent, index, max_length, n_per_col, pal, palettes, radius, svg, text, textwidth, vscale, vscaleL, vscaleR, width;

height = 400;

width = 800;

indent = 15;

textwidth = 125;

radius = 15;

gap_betw_cols = 30;

svg = d3.select("div#chart").append("svg").attr("height", height).attr("width", width);

bgrect = svg.append("rect").attr("x", 2).attr("y", 2).attr("height", height - 4).attr("width", width - 54).attr("fill", "white").attr("stroke", "black").attr("stroke-width", "3").on("click", function(d) {
  bgrect.attr("fill", "white");
  return text.attr("fill", "#333");
});

palettes = Object.keys(wes_palettes);

// vscale to make two columns
n_per_col = Math.ceil(palettes.length / 2);

// vertical scale (in two columns)
vscaleL = d3.scalePoint().domain(d3.range(n_per_col)).range([0, height]).padding(1);

vscaleR = d3.scalePoint().domain(d3.range(n_per_col).map(function(d) {
  return d + n_per_col;
})).range([0, height]).padding(1);

vscale = function(d) {
  if (d < n_per_col) {
    return vscaleL(d);
  }
  return vscaleR(d);
};

// horizontal offset (in two columns)
hoffset = function(d) {
  if (d < n_per_col) {
    return indent;
  }
  return indent + width / 2 + gap_betw_cols / 2;
};

text = svg.selectAll("empty").data(palettes).enter().append("text").text(function(d) {
  return d;
}).attr("x", function(d, i) {
  return hoffset(i);
}).attr("y", function(d, i) {
  return vscale(i);
}).attr("fill", "#333");

max_length = d3.max(palettes.map(function(pal) {
  return wes_palettes[pal].length;
}));

hscale = d3.scalePoint().domain(d3.range(max_length)).range([textwidth, width / 2 - gap_betw_cols / 2]).padding(1);

dark_colors = {
  BottleRocket: [4, 5, 6],
  Cavalcanti: [1],
  Darjeeling2: [4],
  GrandBudapest: [2],
  Moonrise1: [3],
  Moonrise2: [3],
  Rushmore: [3]
};

has_dark = Object.keys(dark_colors);

for (index in palettes) {
  pal = palettes[index];
  svg.append("g").attr("id", pal).selectAll("empty").data(wes_palettes[pal]).enter().append("circle").attr("cy", vscale(index)).attr("cx", function(d, i) {
    return hoffset(index) + hscale(i);
  }).attr("r", radius).attr("fill", function(d) {
    return d;
  }).attr("class", function(d, i) {
    if (has_dark.indexOf(pal) >= 0 && dark_colors[pal].indexOf(i) >= 0) {
      return "dark";
    }
    return "light";
  }).on("click", function(d, i) {
    var f, this_class;
    bgrect.attr("fill", d);
    this_class = d3.select(this).attr("class");
    f = function() {
      if (this_class === "dark") {
        return "#ddd";
      }
      return "#333";
    };
    return text.attr("fill", f);
  });
}


wes_scales.js
// d3 v4 scales for wes anderson palettes
d3.scaleBottleRocket1=function(){return d3.scaleOrdinal().range(["#A42820","#5F5647","#9B110E","#3F5151","#4E2A1E","#550307","#0C1707"]);}
d3.scaleBottleRocket2=function(){return d3.scaleOrdinal().range(["#FAD510","#CB2314","#273046","#354823","#1E1E1E"]);}
d3.scaleCavalcanti1=function(){return d3.scaleOrdinal().range(["#D8B70A","#02401B","#A2A475","#81A88D","#972D15"]);}
d3.scaleChevalier1=function(){return d3.scaleOrdinal().range(["#446455","#FDD262","#D3DDDC","#C7B19C"]);}
d3.scaleDarjeeling1=function(){return d3.scaleOrdinal().range(["#FF0000","#00A08A","#F2AD00","#F98400","#5BBCD6"]);}
d3.scaleDarjeeling2=function(){return d3.scaleOrdinal().range(["#ECCBAE","#046C9A","#D69C4E","#ABDDDE","#000000"]);}
d3.scaleFantasticFox1=function(){return d3.scaleOrdinal().range(["#DD8D29","#E2D200","#46ACC8","#E58601","#B40F20"]);}
d3.scaleGrandBudapest1=function(){return d3.scaleOrdinal().range(["#F1BB7B","#FD6467","#5B1A18","#D67236"]);}
d3.scaleGrandBudapest2=function(){return d3.scaleOrdinal().range(["#E6A0C4","#C6CDF7","#D8A499","#7294D4"]);}
d3.scaleMoonrise1=function(){return d3.scaleOrdinal().range(["#F3DF6C","#CEAB07","#D5D5D3","#24281A"]);}
d3.scaleMoonrise2=function(){return d3.scaleOrdinal().range(["#798E87","#C27D38","#CCC591","#29211F"]);}
d3.scaleMoonrise3=function(){return d3.scaleOrdinal().range(["#85D4E3","#F4B5BD","#9C964A","#CDC08C","#FAD77B"]);}
d3.scaleRoyal1=function(){return d3.scaleOrdinal().range(["#899DA4","#C93312","#FAEFD1","#DC863B"]);}
d3.scaleRoyal2=function(){return d3.scaleOrdinal().range(["#9A8822","#F5CDB4","#F8AFA8","#FDDDA0","#74A089"]);}
d3.scaleRushmore1=function(){return d3.scaleOrdinal().range(["#E1BD6D","#EABE94","#0B775E","#35274A","#F2300F"]);}
d3.scaleZissou1=function(){return d3.scaleOrdinal().range(["#3B9AB2","#78B7C5","#EBCC2A","#E1AF00","#F21A00"]);}


src/create_wesandersonCSS.R
# grab palettes from wesanderson R package and make CSS file
# (https://github.com/karthik/wesanderson)

if(!require(wesanderson)) {
    library(devtools)
    install_github("karthik/wesanderson")
    library(wesanderson)
}

# sort by name
palnames <- sort(names(wes_palettes))

##############################
# create wes_palettes.css
##############################
file <- "../wesanderson.css"
cat("// palettes from https://github.com/karthik/wesanderson\n",
    file=file)
for(pal in palnames) {
    val <- col2rgb(wes_palettes[[pal]])
    class_name <- paste0(".", pal, ".q", (1:ncol(val))-1)

    css <- apply(val, 2, function(a)
                 paste0("{fill:rgb(", paste(a, collapse=","), ")}"))
    for(i in seq(along=css))
        cat(class_name[i], css[i], "\n", sep="", file=file, append=TRUE)
}


wesanderson.js
// palettes from https://github.com/karthik/wesanderson
var wes_palettes = {"BottleRocket1":["#A42820","#5F5647","#9B110E","#3F5151","#4E2A1E","#550307","#0C1707"],"BottleRocket2":["#FAD510","#CB2314","#273046","#354823","#1E1E1E"],"Cavalcanti1":["#D8B70A","#02401B","#A2A475","#81A88D","#972D15"],"Chevalier1":["#446455","#FDD262","#D3DDDC","#C7B19C"],"Darjeeling1":["#FF0000","#00A08A","#F2AD00","#F98400","#5BBCD6"],"Darjeeling2":["#ECCBAE","#046C9A","#D69C4E","#ABDDDE","#000000"],"FantasticFox1":["#DD8D29","#E2D200","#46ACC8","#E58601","#B40F20"],"GrandBudapest1":["#F1BB7B","#FD6467","#5B1A18","#D67236"],"GrandBudapest2":["#E6A0C4","#C6CDF7","#D8A499","#7294D4"],"IsleofDogs1":["#9986A5","#79402E","#CCBA72","#0F0D0E","#D9D0D3","#8D8680"],"IsleofDogs2":["#EAD3BF","#AA9486","#B6854D","#39312F","#1C1718"],"Moonrise1":["#F3DF6C","#CEAB07","#D5D5D3","#24281A"],"Moonrise2":["#798E87","#C27D38","#CCC591","#29211F"],"Moonrise3":["#85D4E3","#F4B5BD","#9C964A","#CDC08C","#FAD77B"],"Royal1":["#899DA4","#C93312","#FAEFD1","#DC863B"],"Royal2":["#9A8822","#F5CDB4","#F8AFA8","#FDDDA0","#74A089"],"Rushmore":["#E1BD6D","#EABE94","#0B775E","#35274A","#F2300F"],"Rushmore1":["#E1BD6D","#EABE94","#0B775E","#35274A","#F2300F"],"Zissou1":["#3B9AB2","#78B7C5","#EBCC2A","#E1AF00","#F21A00"]};
// helper function: call with a palette name and (optionally) some number of colors
var wes_palette = function(palette, n_colors) {
    var palette_names = Object.keys(wes_palettes);
    if(palette_names.indexOf(palette) < 0) {
        console.log("palette " + palette + " not found");
        return null;
    }
    var pal = wes_palettes[palette];
    if(n_colors === null) {
        return pal;
    }
    if(+n_colors === 0) {
        console.log("No colors? That's just silly.");
        return pal;
    }
    if(n_colors > pal.length) {
        console.log("palette " + palette + " only has " + pal.length.toString() + " colors");
        return pal;
    }
    return pal.slice(0,n_colors);
};
